sdk_refactor_design_proposal.md

```diff
@@ -0,0 +1,392 @@
+
+This functionality will be implemented using the [Factory pattern](https://refactoring.guru/design-patterns/factory-method). Each factory-generated class will have its own implementation and validation mechanism. This will make the implementation easy to maintain as this makes it easy to swap algorithms used inside an object during runtime. That way, we can isolate the implementation details from the *Client*. If we choose to implement a new strategy, for example fetching values from environment values, this will not require a change in the Client.
+
+##### Proposed flow
+
+```python
+config = client.Configuration(action=file)
+creds = client.Credental(action=keystore)
+client. conjur.ConjurClient(config, creds, debug, ssl_verify)
+```
+
+```python
+# data object representing the configuration data 
+# needed to connect to the Conjur server
+class ConfigurationModel(action = None,
+                         url = None,
+                         account = None,
+                         ca_bundle = None,
+                         ssl_verify = True):
+
+# configuration factory deciding which flow to run based on user input
+class Configuration(action=None, 
+                    ssl_verify=None)
+
+	if configuration.action == "file":
+		return ConfigurationFromFile(configuration)
+  elif configuration.action == "env":
+		return ConfigurationFromEnv(configuration)
+  
+class ConfigurationFromFile(ConfigurationStrategy)
+	def __init__(url=None, 
+               account=None,
+               ca_bundle=None,
+               ssl_verify=True)
+		
+    self.validate()
+    
+  # validates all parameters were given
+  def validate():
+    if not self.url:
+      raise ValueError("Error: Appliance URL not provided") 
+    # ...
+```
+
+Note: the same pattern will be implemented for Credentials. 
+
+For both Configuration and Credential flows, we will still support the ability for users to build these objects themselves. If so, we won’t fetch the values from disk or any other supported mediums for them. For example:
+
+```python
+config = client.Configuration(url="https://someserver", account="cucumber", cert_file="/some/certfile")
+creds = client.Credental(machine="https://someserver", username="someuser", api_key="123")
+client. conjur.ConjurClient(config, creds, debug, ssl_verify)
+```
+
+We will *not* create the conjurrc and the netrc for the user as is true in the current state of the SDK. 
+
+We will *not* support partial states as is true in the current SDK. For example, we will not support the flow if a user provides `client.Configuration(action=file, url="https://someserver", account="", cert_file="/some/certfile")`. In this flow we will fail with relevant error.
+
+
+
+##### API and commands
+
+The following code excerpt was taken from the current implementation of *Client* and *API* classes:
```

Comment:
regarding the API I suggest going with different approach. implement webapi service as done in other clientside applications.
something like 
`class RestAipProvider:

    def __init__(self, url, api_key=None, ssl_verify=None):
        self.url = url
        self.api_key = api_key
        self.ssl_verify = ssl_verify

    async def authenticate(self, account, login, pasrams=None):

        endpoint = f"{self.url}/authn/{account}/{login}/authenticate"
        async with aiohttp.ClientSession() as session:
            async with session.post(endpoint, data=pasrams) as resp:
                return await resp.json()


    async def get_variable(self, account, kind, identifier, params=None):
        endpoint = f"{self.url}/secrets/{account}/{kind}/{identifier}"
        async with aiohttp.ClientSession() as session:
            async with session.get(endpoint, data=params) as resp:
                await resp.json()
        return resp`
Where each function exposes serverside endpoint and return a promise with the server data response.
This layer should not try to transform the data
The benefits for this approach
1. webapi is scoped to one layer
2. very easy to extand\ add new apis
3. short and very easy to test functions
4. 100% async (we don't have to but I think we do specially for the sdk)

Reply:
The downside that it violates SRP

Reply:
Note that in this method you are actually creating a [facade](https://refactoring.guru/design-patterns/facade). Not saying this make this idea not good but it also means you are leaking the other system terminology (conjur server) into your application. The up side IMO is clear understanding of the calls we make

Reply:
We don't have to use the same terminology..
The function in this API knows the server endpoint but exposes it own functionality. I agree that in most of the cases it will be straightForward but it will be our choice to do so. BTW I don't think it's validate the SRP as the server and the CLI are two different programs. 
Another thing: working with tools like protobuff or bond would have auto generate this layer for you, making me believe this is a step to the right direction